sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/AbstractFixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.exemplar;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.metrics.data.DoubleExemplar;
+import io.opentelemetry.sdk.metrics.data.Exemplar;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Base implementation for fixed-size reservoir sampling of Exemplars.
+ *
+ * <p>Additionally this implementation ONLY exports double valued exemplars.
+ */
+abstract class AbstractFixedSizeExemplarReservoir implements ExemplarReservoir {
+  private final Clock clock;
+  private final ReservoirCell[] storage;
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   */
+  AbstractFixedSizeExemplarReservoir(Clock clock, int size) {
+    this.clock = clock;
+    this.storage = new ReservoirCell[size];
+    for (int i = 0; i < size; ++i) {
+      this.storage[i] = new ReservoirCell();
+    }
+  }
+
+  protected final int maxSize() {
+    return storage.length;
+  }
+  /**
+   * Determines the sample bucket for a given measurement.
+   *
+   * @return The bucket to sample into or -1 for no sampling.
+   */
+  protected abstract int bucketFor(double value, Attributes attributes, Context context);
+  /** Callback to reset any local state after a {@link #collectAndReset} call. */
+  protected void reset() {}
+
+  @Override
+  public final void offerMeasurement(long value, Attributes attributes, Context context) {
+    offerMeasurement((double) value, attributes, context);
+  }
+
+  @Override
+  public final void offerMeasurement(double value, Attributes attributes, Context context) {
+    int bucket = bucketFor(value, attributes, context);
+    if (bucket != -1) {
+      this.storage[bucket].offerMeasurement(value, attributes, context);
+    }
+  }
+
+  @Override
+  public final List<Exemplar> collectAndReset(Attributes pointAttributes) {
+    // Note: we are collecting exemplars from buckets piecemeal, but we
+    // could still be sampling exemplars during this process.
+    List<Exemplar> results = new ArrayList<>();
```

Comment:
```suggestion
    List<Exemplar> results = new ArrayList<>(storage.length);
```

I guess it's common to have a full reservoir?

Reply:
Unclear, actually.  For histograms I'd argue it's NOT common given our fixed size bucketing strategy.

Can we defer this for some sort of profiling?